src/transaction/log_replication_atomic.cpp

```diff
@@ -0,0 +1,173 @@
+ *  limitations under the License.
+ *
+ */
+
+#include "atomic_replicator.hpp"
+
+#include "log_recovery_redo_parallel.hpp"
+
+namespace cublog
+{
+
+  void atomic_replicator::redo_upto (cubthread::entry &thread_entry, const log_lsa &end_redo_lsa)
+  {
+    assert (m_redo_lsa < end_redo_lsa);
+
+    // redo all records from current position (m_redo_lsa) until end_redo_lsa
+
+    m_perfmon_redo_sync.start ();
+    // make sure the log page is refreshed. otherwise it may be outdated and new records may be missed
+    (void) m_redo_context.m_reader.set_lsa_and_fetch_page (m_redo_lsa, log_reader::fetch_mode::FORCE);
+
+    while (m_redo_lsa < end_redo_lsa)
+      {
+	// read and redo a record
+	(void) m_redo_context.m_reader.set_lsa_and_fetch_page (m_redo_lsa);
+
+	const log_rec_header header = m_redo_context.m_reader.reinterpret_copy_and_add_align<log_rec_header> ();
+
+	switch (header.type)
+	  {
+	  case LOG_REDO_DATA:
+	    read_and_redo_record_with_atomic_consideration<log_rec_redo> (thread_entry, header.type, header.back_lsa, m_redo_lsa,
+		header.trid);
+	    break;
+	  case LOG_MVCC_REDO_DATA:
+	    read_and_redo_record_with_atomic_consideration<log_rec_mvcc_redo> (thread_entry, header.type, header.back_lsa,
+		m_redo_lsa, header.trid);
+	    break;
+	  case LOG_UNDOREDO_DATA:
+	  case LOG_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record_with_atomic_consideration<log_rec_undoredo> (thread_entry, header.type, header.back_lsa,
+		m_redo_lsa, header.trid);
+	    break;
+	  case LOG_MVCC_UNDOREDO_DATA:
+	  case LOG_MVCC_DIFF_UNDOREDO_DATA:
+	    read_and_redo_record_with_atomic_consideration<log_rec_mvcc_undoredo> (thread_entry, header.type, header.back_lsa,
+		m_redo_lsa, header.trid);
+	    break;
+	  case LOG_RUN_POSTPONE:
+	    read_and_redo_record_with_atomic_consideration<log_rec_run_postpone> (thread_entry, header.type, header.back_lsa,
+		m_redo_lsa, header.trid);
+	    break;
+	  case LOG_COMPENSATE:
+	    read_and_redo_record_with_atomic_consideration<log_rec_compensate> (thread_entry, header.type, header.back_lsa,
+		m_redo_lsa, header.trid);
+	    break;
+	  case LOG_DBEXTERN_REDO_DATA:
+	  {
+	    const log_rec_dbout_redo dbout_redo =
+		    m_redo_context.m_reader.reinterpret_copy_and_add_align<log_rec_dbout_redo> ();
+	    log_rcv rcv;
+	    rcv.length = dbout_redo.length;
+
+	    log_rv_redo_record (&thread_entry, m_redo_context.m_reader, RV_fun[dbout_redo.rcvindex].redofun, &rcv,
+				&m_redo_lsa, 0, nullptr, m_redo_context.m_redo_zip);
+	    break;
+	  }
+	  case LOG_COMMIT:
+	    calculate_replication_delay_or_dispatch_async<log_rec_donetime> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_ABORT:
+	    calculate_replication_delay_or_dispatch_async<log_rec_donetime> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_DUMMY_HA_SERVER_STATE:
+	    calculate_replication_delay_or_dispatch_async<log_rec_ha_server_state> (
+		    thread_entry, m_redo_lsa);
+	    break;
+	  case LOG_TRANTABLE_SNAPSHOT:
```

Comment:
`LOG_TRANTABLE_SNAPSHOT` does not need to be handled on PTS; replace handling with a comment; thus `m_most_recent_trantable_snapshot_lsa` can be private in base class

Reply:
I am confused about the comment on the replicator:
```
	    // save the LSA of the last transaction table snapshot that can be found in the log
	    // only needed on the passive transaction server
```

Reply:
Yes, the comment is somewhat contrived. "Needed" refers to the final usage which is indeed on the PTS.
It should read:
```
// save the LSA of the last transaction table snapshot that can be found in the log
// transaction table snapshots are added to the transactional log by the active transaction server
// the LSA of the most recent is saved/bookkept by the page server (this section)
// and, finally, this LSA is retrieved and used by a booting up passive transaction server
```

Reply:
Please change the comment in `log_replication::redo_upto`.